File: ObjectiveC/CollectionView101/CollectionView101/GroupingSampleTableViewController.m
//
//  GroupingSampleTableViewController.m
//  CollectionView101
//
//  Copyright © 2016 GrapeCity. All rights reserved.
//

#import "GroupingSampleTableViewController.h"
#import "YouTubeCollectionView.h"
#import "SortingSampleTableViewController.h"
#import "SampleData.h"


@interface GroupingSampleTableViewController ()
@property YouTubeCollectionView* collectionView;
@end

@implementation GroupingSampleTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _collectionView = [[YouTubeCollectionView alloc] initWithQuery:@"xamarin" orderBy:@"relevance"];
    [_collectionView loadMoreItems:@(50) completion:^{
        XuniPropertyGroupDescription *gd = [[XuniPropertyGroupDescription alloc] initWithProperty:@"channelTitle"];
        [_collectionView.groupDescriptions addObject:gd];
        [self.tableView reloadData];
    }];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.collectionView.groups.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    XuniCollectionViewGroup* group = [self.collectionView.groups objectAtIndex:section];
    return group.items.count;
}

-(NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    XuniCollectionViewGroup* group = [self.collectionView.groups objectAtIndex:section];
    SampleData* thisData = (SampleData*)[group.items objectAtIndex:0];
    return thisData.channelTitle;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"YouTubeEntryCell" forIndexPath:indexPath];
    
    XuniCollectionViewGroup* group = [self.collectionView.groups objectAtIndex:indexPath.section];
    
    SampleData* thisData = (SampleData*)[group.items objectAtIndex:indexPath.row];
    cell.textLabel.text = thisData.title;
    cell.detailTextLabel.text = thisData.subtitle;
    
    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSData* data = [NSData dataWithContentsOfURL:[NSURL URLWithString:thisData.thumbnail]];
        dispatch_async(dispatch_get_main_queue(), ^{
            cell.imageView.image = [UIImage imageWithData:data];
            [cell layoutSubviews];
        });
    });
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    XuniCollectionViewGroup* group = [self.collectionView.groups objectAtIndex:indexPath.section];
    
    SampleData* thisData = (SampleData*)[group.items objectAtIndex:indexPath.row];
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:thisData.link]];
}


@end

File: Swift/CollectionView101/CollectionView101/GroupingSampleTableViewController.swift
//
//  GroupingSampleTableViewController.h
//  CollectionView101
//
//  Copyright © 2016 GrapeCity. All rights reserved.
//
import UIKit
import XuniCoreKit

class GroupingSampleTableViewController: UITableViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.collectionView = YouTubeCollectionView(query: "xamarin", orderBy: "relevance")
        collectionView.loadMoreItems(50, completion: {() -> Void in
            let gd: XuniPropertyGroupDescription = XuniPropertyGroupDescription(property: "channelTitle")
            self.collectionView.groupDescriptions.addObject(gd)
            self.tableView.reloadData()
        })
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return self.collectionView.groups.count
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let group: XuniCollectionViewGroup = self.collectionView.groups[section] as! XuniCollectionViewGroup
        return group.items.count
    }

    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String {
        let group: XuniCollectionViewGroup = self.collectionView.groups[section] as! XuniCollectionViewGroup
        let thisData: SampleData = (group.items[0] as! SampleData)
        return thisData.channelTitle
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell: UITableViewCell = tableView.dequeueReusableCellWithIdentifier("YouTubeEntryCell", forIndexPath: indexPath)
        let thisData: SampleData = (self.collectionView.items[indexPath.row] as! SampleData)
        cell.textLabel!.text = thisData.title
        cell.detailTextLabel!.text = thisData.subtitle
        
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {() -> Void in
            if let data = NSData(contentsOfURL:NSURL(string: thisData.thumbnail)!) {
                dispatch_async(dispatch_get_main_queue(), {() -> Void in
                    cell.imageView?.image = UIImage(data: data)
                    cell.layoutSubviews()
                })
            }
        })
        return cell
    }

    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let group: XuniCollectionViewGroup = self.collectionView.groups[indexPath.section] as! XuniCollectionViewGroup
        let thisData: SampleData = (group.items[indexPath.row] as! SampleData)
        UIApplication.sharedApplication().openURL(NSURL(string: thisData.link)!)
    }

    var collectionView: YouTubeCollectionView!
}